src/python/pants/backend/codegen/protobuf/scala/rules.py

```diff
@@ -0,0 +1,181 @@
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+import textwrap
+
+from pants.backend.codegen.protobuf.protoc import Protoc
+from pants.backend.codegen.protobuf.scala.scalapbc import ScalaPBSubsystem
+from pants.backend.codegen.protobuf.target_types import (
+    ProtobufDependenciesField,
+    ProtobufSourceField,
+)
+from pants.backend.scala.target_types import ScalaSourceField
+from pants.core.util_rules.external_tool import DownloadedExternalTool, ExternalToolRequest
+from pants.core.util_rules.source_files import SourceFilesRequest
+from pants.core.util_rules.stripped_source_files import StrippedSourceFiles
+from pants.engine.addresses import Addresses, UnparsedAddressInputs
+from pants.engine.fs import (
+    AddPrefix,
+    CreateDigest,
+    Digest,
+    Directory,
+    FileContent,
+    MergeDigests,
+    RemovePrefix,
+    Snapshot,
+)
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.platform import Platform
+from pants.engine.process import BashBinary, Process, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import (
+    GeneratedSources,
+    GenerateSourcesRequest,
+    InjectDependenciesRequest,
+    InjectedDependencies,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.jvm.jdk_rules import JdkSetup
+from pants.jvm.resolve.coursier_fetch import MaterializedClasspath, MaterializedClasspathRequest
+from pants.jvm.resolve.jvm_tool import JvmToolLockfileRequest, JvmToolLockfileSentinel
+from pants.source.source_root import SourceRoot, SourceRootRequest
+from pants.util.logging import LogLevel
+
+
+class GenerateScalaFromProtobufRequest(GenerateSourcesRequest):
+    input = ProtobufSourceField
+    output = ScalaSourceField
+
+
+class ScalapbcToolLockfileSentinel(JvmToolLockfileSentinel):
+    options_scope = ScalaPBSubsystem.options_scope
+
+
+@rule(desc="Generate Scala from Protobuf", level=LogLevel.DEBUG)
+async def generate_scala_from_protobuf(
+    request: GenerateScalaFromProtobufRequest,
+    protoc: Protoc,
+    scalapbc: ScalaPBSubsystem,
+    jdk_setup: JdkSetup,
+    bash: BashBinary,
+) -> GeneratedSources:
+    output_dir = "_generated_files"
+
+    downloaded_protoc_binary, tool_classpath, empty_output_dir, transitive_targets = await MultiGet(
+        Get(DownloadedExternalTool, ExternalToolRequest, protoc.get_request(Platform.current)),
+        Get(
+            MaterializedClasspath,
+            MaterializedClasspathRequest(
+                prefix="__toolcp",
+                lockfiles=(scalapbc.resolved_lockfile(),),
+            ),
+        ),
+        Get(Digest, CreateDigest([Directory(output_dir)])),
+        Get(TransitiveTargets, TransitiveTargetsRequest([request.protocol_target.address])),
+    )
+
+    # NB: By stripping the source roots, we avoid having to set the value `--proto_path`
+    # for Protobuf imports to be discoverable.
```

Comment:
Might be good to prefix them after stripping, to avoid collisions? But maybe not a problem.

Reply:
The note and this code comes directly from the Python protobuf rules. I don't have enough context to know if collisions would be a problem. Both Python and Scala rules should probably compute the correct search path and not strip source roots. But that can be a separate PR since we'd need to fix both places.